File: server/sync-server.js
const express = require("express");
const cors = require("cors");
const multer = require("multer");
const path = require("path");
const fs = require("fs");

const app = express();
const PORT = 3001;

// 确保上传目录存在
const UPLOAD_DIR = path.join(__dirname, "uploads");
if (!fs.existsSync(UPLOAD_DIR)) {
    fs.mkdirSync(UPLOAD_DIR, { recursive: true });
}

// 配置 multer 存储
const storage = multer.diskStorage({
    destination: (req, file, cb) => {
        cb(null, UPLOAD_DIR);
    },
    filename: (req, file, cb) => {
        const uniqueSuffix = Date.now() + "-" + Math.round(Math.random() * 1e9);
        const ext = path.extname(file.originalname);
        cb(null, uniqueSuffix + ext);
    },
});

const upload = multer({
    storage,
    limits: { fileSize: 50 * 1024 * 1024 }, // 50MB
    fileFilter: (req, file, cb) => {
        const allowedTypes = /jpeg|jpg|png|gif|webp|svg/;
        const ext = allowedTypes.test(path.extname(file.originalname).toLowerCase());
        const mime = allowedTypes.test(file.mimetype);
        if (ext && mime) {
            cb(null, true);
        } else {
            cb(new Error("只支持图片文件"));
        }
    },
});

// 中间件
app.use(cors());
app.use(express.json({ limit: "50mb" }));

// 静态文件服务 - 提供上传的图片访问
app.use("/uploads", express.static(UPLOAD_DIR));

// ─── 数据存储（内存） ─────────────────────────────────────────

/** 全局序号计数器 */
let seqCounter = 0;

/** 事件数组 */
let events = [];

/** 画布全量数据 */
let canvasJSON = null;

/** SSE 客户端连接列表 */
const sseClients = new Map();

// ─── SSE 广播 ─────────────────────────────────────────

/**
 * 向所有客户端广播事件
 */
function broadcast(event) {
    const message = JSON.stringify({
        type: "sync_event",
        data: event,
    });

    sseClients.forEach((res, clientId) => {
        res.write(`data: ${message}\n\n`);
    });

    console.log(`[广播] seq=${event.seq}, eventType=${event.eventType}, 客户端数=${sseClients.size}`);
}


// ─── 接口 ─────────────────────────────────────────

/**
 * SSE 连接端点
 * GET /api/canvas/sync/sse?clientId=xxx
 */
app.get("/api/canvas/sync/sse", (req, res) => {
    const clientId = req.query.clientId;
    if (!clientId) {
        return res.status(400).json({ error: "缺少 clientId" });
    }

    // 设置 SSE 响应头
    res.setHeader("Content-Type", "text/event-stream");
    res.setHeader("Cache-Control", "no-cache");
    res.setHeader("Connection", "keep-alive");
    res.setHeader("Access-Control-Allow-Origin", "*");

    // 发送初始连接成功消息
    res.write(`data: ${JSON.stringify({ type: "connected", clientId })}\n\n`);

    // 注册客户端
    sseClients.set(clientId, res);
    console.log(`[SSE] 客户端连接: ${clientId}, 当前连接数: ${sseClients.size}`);

    // 客户端断开时清理
    req.on("close", () => {
        sseClients.delete(clientId);
        console.log(`[SSE] 客户端断开: ${clientId}, 当前连接数: ${sseClients.size}`);
    });
});

/**
 * 推送同步事件
 * POST /api/canvas/sync/event
 */
app.post("/api/canvas/sync/event", (req, res) => {
    const { eventType, data } = req.body;

    if (!eventType || !data) {
        return res.status(400).json({ error: "缺少 eventType 或 data" });
    }

    // 分配全局序号
    seqCounter++;
    const event = {
        seq: seqCounter,
        eventType,
        data,
    };

    // 存入事件数组
    events.push(event);

    console.log(`[事件] seq=${event.seq}, eventType=${eventType}, 事件总数=${events.length}`);

    // 广播给所有客户端
    broadcast(event);

    // 返回响应
    res.json({
        seq: event.seq,
        eventArrayLength: events.length,
    });
});

/**
 * 上传全量画布数据
 * POST /api/canvas/sync/full
 */
app.post("/api/canvas/sync/full", (req, res) => {
    const { clientId, canvasJSON: newCanvasJSON } = req.body;

    if (!clientId || !newCanvasJSON) {
        return res.status(400).json({ error: "缺少 clientId 或 canvasJSON" });
    }

    // 存储全量数据
    canvasJSON = newCanvasJSON;

    // 清空事件数组
    events = [];

    console.log(`[全量同步] clientId=${clientId}, 事件数组已清空`);

    res.json({ success: true });
});

/**
 * 获取初始化数据
 * GET /api/canvas/sync/full_data
 */
app.get("/api/canvas/sync/full_data", (req, res) => {
    console.log(`[初始化] 返回全量数据和 ${events.length} 个事件`);

    res.json({
        canvasJSON,
        events,
    });
});

/**
 * 调试接口：查看当前状态
 * GET /api/canvas/sync/debug
 */
app.get("/api/canvas/sync/debug", (req, res) => {
    res.json({
        seqCounter,
        eventsCount: events.length,
        events: events.slice(-10), // 只返回最近10条
        hasCanvasJSON: canvasJSON !== null,
        connectedClients: Array.from(sseClients.keys()),
    });
});

/**
 * 调试接口：重置所有数据
 * POST /api/canvas/sync/reset
 */
app.post("/api/canvas/sync/reset", (req, res) => {
    seqCounter = 0;
    events = [];
    canvasJSON = null;
    console.log("[重置] 所有数据已清空");
    res.json({ success: true });
});


/**
 * 调试接口：后端主动插入图片
 * POST /api/canvas/sync/inject_image
 * 后端只负责传递图片 URL，前端负责封装为 HistoryRecord
 */
app.post("/api/canvas/sync/inject_image", (req, res) => {
    let { urls } = req.body;

    // 如果没有传 urls，从 uploads 目录随机选一张图片
    if (!urls || urls.length === 0) {
        const files = fs.readdirSync(UPLOAD_DIR).filter((f) =>
            /\.(jpg|jpeg|png|gif|webp)$/i.test(f)
        );

        if (files.length === 0) {
            return res.status(400).json({ error: "uploads 目录下没有图片，且未提供 urls 参数" });
        }

        const randomFile = files[Math.floor(Math.random() * files.length)];
        urls = [`http://localhost:${PORT}/uploads/${randomFile}`];
    }

    // 构造 server:add_image 事件
    seqCounter++;
    const event = {
        seq: seqCounter,
        eventType: "server:add_image",
        data: {
            urls,
        },
    };
    events.push(event);

    // 广播给所有客户端
    broadcast(event);

    console.log(`[注入图片] seq=${event.seq}, urls=${urls}`);

    res.json({
        success: true,
        seq: event.seq,
        urls,
    });
});

// ─── 图片上传接口 ─────────────────────────────────────────

/**
 * 上传图片
 * POST /api/upload/image
 * 返回图片的访问 URL
 */
app.post("/api/upload/image", upload.single("image"), (req, res) => {
    if (!req.file) {
        return res.status(400).json({ error: "没有上传文件" });
    }

    const url = `http://localhost:${PORT}/uploads/${req.file.filename}`;
    console.log(`[上传] 图片已保存: ${req.file.filename}`);

    res.json({
        success: true,
        url,
        filename: req.file.filename,
        originalName: req.file.originalname,
        size: req.file.size,
    });
});

// ─── 启动服务器 ─────────────────────────────────────────

app.listen(PORT, () => {
    console.log(`\n========================================`);
    console.log(`  画布同步服务器已启动`);
    console.log(`  地址: http://localhost:${PORT}`);
    console.log(`========================================\n`);
    console.log(`接口列表:`);
    console.log(`  GET  /api/canvas/sync/sse?clientId=xxx  - SSE 连接`);
    console.log(`  POST /api/canvas/sync/event             - 推送事件`);
    console.log(`  POST /api/canvas/sync/full              - 全量同步`);
    console.log(`  GET  /api/canvas/sync/full_data         - 获取初始化数据`);
    console.log(`  GET  /api/canvas/sync/debug             - 调试状态`);
    console.log(`  POST /api/canvas/sync/reset             - 重置数据`);
    console.log(`  POST /api/canvas/sync/inject_image      - 后端注入图片`);
    console.log(`  POST /api/upload/image                  - 上传图片`);
    console.log(`  GET  /uploads/:filename                 - 访问图片\n`);
});

File: server/sync_server.py
import os
import random
import asyncio
from datetime import datetime
from typing import Optional, Dict, Any, List, Literal, Union
from pathlib import Path

from fastapi import FastAPI, File, UploadFile, HTTPException, Query
from fastapi.middleware.cors import CORSMiddleware
from fastapi.staticfiles import StaticFiles
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
import aiofiles
import json

app = FastAPI(title="画布同步服务器")

PORT = 3001

# 确保上传目录存在
UPLOAD_DIR = Path(__file__).parent / "uploads"
UPLOAD_DIR.mkdir(parents=True, exist_ok=True)

# 允许的图片类型
ALLOWED_EXTENSIONS = {".jpeg", ".jpg", ".png", ".gif", ".webp", ".svg"}
MAX_FILE_SIZE = 50 * 1024 * 1024  # 50MB

# CORS 配置
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

# 静态文件服务 - 提供上传的图片访问
app.mount("/uploads", StaticFiles(directory=str(UPLOAD_DIR)), name="uploads")

# ─── 数据存储（内存） ─────────────────────────────────────────

sequence_id = 0  # 自增序列号
canvas_data: List[Dict[str, Any]] = []  # 画布数据，ImageExportData 数组
sse_clients: Dict[str, asyncio.Queue] = {}

# ─── Pydantic 模型 ─────────────────────────────────────────

class SyncEventItem(BaseModel):
    """单个同步事件项"""
    event_type: Literal["client:change"]
    event_data: Dict[str, Any]  # ImageExportData 格式

class SyncPushRequest(BaseModel):
    """推送同步事件的请求体"""
    events: List[SyncEventItem]

class FullSyncRequest(BaseModel):
    """全量同步请求"""
    clientId: str
    canvasJSON: List[Dict[str, Any]]  # ImageExportData 数组

class InjectImageRequest(BaseModel):
    urls: Optional[List[str]] = None

# ─── 画布数据操作 ─────────────────────────────────────────

def merge_event_to_canvas(event_data: Dict[str, Any]) -> None:
    """
    将事件数据合并到画布中
    - 按 id 查找，找到则替换，找不到则 push
    """
    global canvas_data
    
    event_id = event_data.get("id")
    if not event_id:
        print(f"[警告] 事件数据缺少 id: {event_data}")
        return
    
    # 查找是否已存在
    found_index = -1
    for i, item in enumerate(canvas_data):
        if item.get("id") == event_id:
            found_index = i
            break
    
    if found_index >= 0:
        # 找到了，替换
        canvas_data[found_index] = event_data
        print(f"[合并] 替换对象 id={event_id}")
    else:
        # 没找到，push
        canvas_data.append(event_data)
        print(f"[合并] 新增对象 id={event_id}")

# ─── SSE 广播 ─────────────────────────────────────────

async def broadcast(event: Dict[str, Any]):
    """向所有客户端广播事件"""
    message = json.dumps({"type": "sync_event", "data": event})
    
    broadcast_count = 0
    for client_id, queue in list(sse_clients.items()):
        try:
            await queue.put(message)
            broadcast_count += 1
        except Exception as e:
            print(f"[广播错误] clientId={client_id}, error={e}")
    
    print(f"[广播] sequence_id={event.get('sequence_id')}, 广播给 {broadcast_count} 个客户端")


# ─── 接口 ─────────────────────────────────────────

@app.get("/api/canvas/sync/sse")
async def sse_endpoint():
    """SSE 连接端点"""
    import uuid
    connection_id = str(uuid.uuid4())[:8]  # 仅用于内部连接管理
    
    async def event_generator():
        queue = asyncio.Queue()
        sse_clients[connection_id] = queue
        print(f"[SSE] 客户端连接: {connection_id}, 当前连接数: {len(sse_clients)}")
        
        try:
            # 发送初始连接成功消息
            yield f"data: {json.dumps({'type': 'connected'})}\n\n"
            
            while True:
                try:
                    message = await asyncio.wait_for(queue.get(), timeout=30)
                    yield f"data: {message}\n\n"
                except asyncio.TimeoutError:
                    # 发送心跳保持连接
                    yield f": heartbeat\n\n"
        except asyncio.CancelledError:
            pass
        finally:
            sse_clients.pop(connection_id, None)
            print(f"[SSE] 客户端断开: {connection_id}, 当前连接数: {len(sse_clients)}")
    
    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "Access-Control-Allow-Origin": "*",
        }
    )


@app.post("/api/canvas/sync/event")
async def push_sync_event(request: SyncPushRequest):
    """
    推送同步事件
    - 接收 events 数组
    - 每个 event 合并到画布数据中
    - sequence_id 自增
    - 先返回响应，再异步广播（让前端更快拿到新 sequence_id）
    """
    global sequence_id
    
    # 收集待广播的事件
    broadcast_events = []
    
    for event_item in request.events:
        # sequence_id 自增
        sequence_id += 1
        
        # 合并到画布数据
        merge_event_to_canvas(event_item.event_data)
        
        # 构造广播事件
        broadcast_events.append({
            "sequence_id": sequence_id,
            "event_type": event_item.event_type,
            "event_data": event_item.event_data,
        })
    
    print(f"[事件] 处理 {len(request.events)} 个事件, sequence_id={sequence_id}, 画布对象数={len(canvas_data)}")
    
    # 异步广播（不阻塞响应）
    asyncio.create_task(broadcast_all(broadcast_events))
    
    return {"sequence_id": sequence_id}


async def broadcast_all(events: List[Dict[str, Any]]):
    """批量广播事件"""
    for event in events:
        await broadcast(event)


@app.post("/api/canvas/sync/full")
async def full_sync(request: FullSyncRequest):
    """上传全量画布数据"""
    global canvas_data
    
    canvas_data = request.canvasJSON
    
    print(f"[全量同步] clientId={request.clientId}, 画布对象数={len(canvas_data)}")
    
    return {"success": True}


@app.get("/api/canvas/sync/full_data")
async def get_full_data():
    """获取初始化数据（画布全量数据）"""
    print(f"[初始化] 返回画布数据, 对象数={len(canvas_data)}")
    
    return {
        "canvasJSON": canvas_data,
        "sequence_id": sequence_id,
    }


@app.get("/api/canvas/sync/debug")
async def debug_status():
    """调试接口：查看当前状态"""
    return {
        "sequence_id": sequence_id,
        "canvasDataCount": len(canvas_data),
        "canvasData": canvas_data,
        "connectedClients": list(sse_clients.keys()),
    }


@app.post("/api/canvas/sync/reset")
async def reset_data():
    """调试接口：重置所有数据"""
    global sequence_id, canvas_data
    
    sequence_id = 0
    canvas_data = []
    
    print("[重置] 所有数据已清空")
    return {"success": True}


@app.post("/api/canvas/sync/inject_image")
async def inject_image(request: InjectImageRequest = InjectImageRequest()):
    """
    调试接口：后端主动插入图片
    构造 ImageExportData 格式的数据并广播
    """
    global sequence_id
    
    urls = request.urls
    
    # 如果没有传 urls 或为空数组，从 uploads 目录随机选一张图片
    if not urls or len(urls) == 0:
        files = [
            f for f in os.listdir(UPLOAD_DIR)
            if Path(f).suffix.lower() in {".jpg", ".jpeg", ".png", ".gif", ".webp"}
        ]
        
        if not files:
            raise HTTPException(status_code=400, detail="uploads 目录下没有图片，且未提供 urls 参数")
        
        random_file = random.choice(files)
        urls = [f"http://localhost:{PORT}/uploads/{random_file}"]
    
    injected_ids = []
    
    for url in urls:
        sequence_id += 1
        
        # 生成唯一 ID
        img_id = f"img_server_{sequence_id}_{random.randint(0, 10**6)}"
        
        # 构造 ImageExportData 格式
        event_data = {
            "id": img_id,
            "metadata": {
                "category": "image",
                "id": img_id,
                "src": url,
                "is_delete": False,
            },
            "style": {
                "matrix": [1, 0, 0, 1, 400, 300],  # 默认位置
            },
        }
        
        # 合并到画布数据
        merge_event_to_canvas(event_data)
        
        # 构造广播事件
        broadcast_event = {
            "sequence_id": sequence_id,
            "event_type": "server:add_image",
            "event_data": event_data,
        }
        
        # 广播给所有客户端
        await broadcast(broadcast_event)
        
        injected_ids.append(img_id)
    
    print(f"[注入图片] 注入 {len(urls)} 张图片, sequence_id={sequence_id}")
    
    return {
        "success": True,
        "sequence_id": sequence_id,
        "ids": injected_ids,
        "urls": urls,
    }


# ─── 图片上传接口 ─────────────────────────────────────────

@app.post("/api/upload/image")
async def upload_image(image: UploadFile = File(...)):
    """上传图片，返回图片的访问 URL"""
    
    # 检查文件类型
    ext = Path(image.filename).suffix.lower()
    if ext not in ALLOWED_EXTENSIONS:
        raise HTTPException(status_code=400, detail="只支持图片文件")
    
    # 生成唯一文件名
    unique_suffix = f"{int(datetime.now().timestamp() * 1000)}-{random.randint(0, 10**9)}"
    filename = f"{unique_suffix}{ext}"
    filepath = UPLOAD_DIR / filename
    
    # 保存文件
    content = await image.read()
    if len(content) > MAX_FILE_SIZE:
        raise HTTPException(status_code=400, detail="文件大小超过 50MB 限制")
    
    async with aiofiles.open(filepath, "wb") as f:
        await f.write(content)
    
    url = f"http://localhost:{PORT}/uploads/{filename}"
    print(f"[上传] 图片已保存: {filename}")
    
    return {
        "success": True,
        "url": url,
        "filename": filename,
        "originalName": image.filename,
        "size": len(content),
    }


# ─── 启动服务器 ─────────────────────────────────────────

if __name__ == "__main__":
    import uvicorn
    
    print(f"\n========================================")
    print(f"  画布同步服务器已启动")
    print(f"  地址: http://localhost:{PORT}")
    print(f"========================================\n")
    print(f"接口列表:")
    print(f"  GET  /api/canvas/sync/sse               - SSE 连接")
    print(f"  POST /api/canvas/sync/event             - 推送事件")
    print(f"  POST /api/canvas/sync/full              - 全量同步")
    print(f"  GET  /api/canvas/sync/full_data         - 获取初始化数据")
    print(f"  GET  /api/canvas/sync/debug             - 调试状态")
    print(f"  POST /api/canvas/sync/reset             - 重置数据")
    print(f"  POST /api/canvas/sync/inject_image      - 后端注入图片")
    print(f"  POST /api/upload/image                  - 上传图片")
    print(f"  GET  /uploads/:filename                 - 访问图片\n")
    
    uvicorn.run(app, host="0.0.0.0", port=PORT)
